src/coreclr/jit/inductionvariableopts.cpp

```diff
@@ -0,0 +1,1519 @@
+        }
+
+    public:
+        bool MadeChanges = false;
+
+        enum
+        {
+            DoPreOrder = true,
+        };
+
+        ReplaceVisitor(Compiler* comp, unsigned lclNum, unsigned ssaNum, unsigned newLclNum)
+            : GenTreeVisitor(comp), m_lclNum(lclNum), m_ssaNum(ssaNum), m_newLclNum(newLclNum)
+        {
+        }
+
+        fgWalkResult PreOrderVisit(GenTree** use, GenTree* user)
+        {
+            GenTree* node = *use;
+            if (node->OperIs(GT_CAST))
+            {
+                GenTreeCast* cast = node->AsCast();
+                if ((cast->gtCastType == TYP_LONG) && cast->IsUnsigned() && !cast->gtOverflow())
+                {
+                    GenTree* op = cast->CastOp();
+                    if (op->OperIs(GT_LCL_VAR) && IsLocal(op->AsLclVarCommon()))
+                    {
+                        *use        = m_compiler->gtNewLclvNode(m_newLclNum, TYP_LONG);
+                        MadeChanges = true;
+                        return fgWalkResult::WALK_SKIP_SUBTREES;
+                    }
+                }
+            }
+            else if (node->OperIs(GT_LCL_VAR, GT_STORE_LCL_VAR, GT_LCL_FLD, GT_STORE_LCL_FLD) &&
+                     IsLocal(node->AsLclVarCommon()))
+            {
+                switch (node->OperGet())
+                {
+                    case GT_LCL_VAR:
+                        node->AsLclVarCommon()->SetLclNum(m_newLclNum);
+                        // No cast needed -- the backend allows TYP_INT uses of TYP_LONG locals.
+                        break;
+                    case GT_STORE_LCL_VAR:
+                    {
+                        node->AsLclVarCommon()->SetLclNum(m_newLclNum);
+                        node->gtType = TYP_LONG;
+                        node->AsLclVarCommon()->Data() =
+                            m_compiler->gtNewCastNode(TYP_LONG, node->AsLclVarCommon()->Data(), true, TYP_LONG);
+                        break;
+                    }
+                    case GT_LCL_FLD:
+                    case GT_STORE_LCL_FLD:
+                        assert(!"Unexpected field use for local not marked as DNER");
+                        break;
+                    default:
+                        break;
+                }
+
+                MadeChanges = true;
+            }
+
+            return fgWalkResult::WALK_CONTINUE;
+        }
+    };
+
+    ReplaceVisitor visitor(this, lclNum, ssaNum, newLclNum);
+    visitor.WalkTree(stmt->GetRootNodePointer(), nullptr);
+    if (visitor.MadeChanges)
+    {
+        gtSetStmtInfo(stmt);
+        fgSetStmtSeq(stmt);
+        JITDUMP("New tree:\n", dspTreeID(stmt->GetRootNode()));
+        DISPTREE(stmt->GetRootNode());
+        JITDUMP("\n");
+    }
+    else
+    {
+        JITDUMP("No replacements made\n");
+    }
+}
+
+//------------------------------------------------------------------------
+// optBestEffortReplaceNarrowIVUsesWith: Try to find and replace uses of the specified
```

Comment:
Renamed it to `optBestEffortReplaceNarrowIVUses`, and added a remarks:
```cpp
// Remarks:
//   This function is best effort; it might not find all uses of the provided
//   SSA num, particularly because it does not follow into joins. Note that we
//   only use this to replace uses of the narrow IV outside the loop; inside
//   the loop we do ensure that all uses/defs are replaced.
//   Keeping it best-effort outside the loop is ok; there is no correctness
//   issue since we do not invalidate the value of the old narrow IV in any
//   way, but it may mean we end up leaving the narrow IV live concurrently
//   with the new widened IV, increasing register pressure.
```